components/movementsensor/gpsrtkserial/gpsrtkserial.go

```diff
@@ -371,47 +390,77 @@ func (g *rtkSerial) closePort() {
 	}
 }
 
-// receiveAndWriteSerial connects to NTRIP receiver and sends correction stream to the MovementSensor through serial.
-func (g *rtkSerial) receiveAndWriteSerial() {
-	defer g.activeBackgroundWorkers.Done()
+// ntripConnection connects to NTRIP stream.
+func (g *rtkSerial) ntripConnection() error {
 	if err := g.cancelCtx.Err(); err != nil {
-		return
+		return g.err.Get()
 	}
+	g.urlMutex.Lock()
 	err := g.connect(g.ntripClient.URL, g.ntripClient.Username, g.ntripClient.Password, g.ntripClient.MaxConnectAttempts)
 	if err != nil {
 		g.err.Set(err)
-		return
+		return g.err.Get()
```

Comment:
If we hit this line, we never unlock the mutex locked on line 398. I'd prefer to always `defer g.urlMutex.Unlock()` immediately after locking to avoid these issues. but I don't have a sense of whether we'll have trouble if the mutex remains locked from lines 419 to the end of the function. If that would cause trouble, maybe we can split the part where it's locked out into a helper function that starts with locking and deferring unlocking the mutex, and then call the helper function from here instead?

Reply:
spoke offline about this. Alan suggested that I should probably make a separate function and use defer on mutex.Unlock().